src/components/application_manager/src/helpers/application_helper.cc

```diff
@@ -9,12 +9,16 @@ namespace {
 using namespace application_manager;
 void DeleteWayPoints(ApplicationSharedPtr app,
                      ApplicationManager& app_manager) {
-  app_manager.UnsubscribeAppFromWayPoints(app);
-  if (!app_manager.IsAnyAppSubscribedForWayPoints()) {
+  std::set<uint32_t> subscribed_apps =
+      app_manager.GetAppsSubscribedForWayPoints();
+  bool send_unsubscribe =
+      subscribed_apps.size() <= 1 && app_manager.IsSubscribedToHMIWayPoints();
```

Comment:
Are unsubscribe requests never sent to mobile app service providers?

Reply:
As of now, correct. Same with subscribe requests. Initially I didn't see a way of doing so using the existing RPC passing handler interface, but it might be possible to do just manually using mobile events (similar to GetAppServiceData), it would just make the logic a lot more complicated. This would also have the counterintuitive consequence that the app service must NOT include `SubscribeWayPoints` or `UnsubscribeWayPoints` (this is still the case with this implementation, but at least the app service provider will never receive these messages), since the RPC passing handler would take over in that case.

I am also not certain how we would handle the case where an app service which handles waypoints registers while another app is already subscribed to waypoints.